crates/sui-graphql-rpc/src/types/transaction_block_kind/programmable.rs

```diff
@@ -0,0 +1,486 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use async_graphql::{
+    connection::{Connection, Edge},
+    *,
+};
+use sui_types::transaction::{
+    Argument as NativeArgument, CallArg as NativeCallArg, Command as NativeProgrammableTransaction,
+    ObjectArg as NativeObjectArg, ProgrammableMoveCall as NativeMoveCallTransaction,
+    ProgrammableTransaction as NativeProgrammableTransactionBlock,
+};
+
+use crate::{
+    context_data::db_data_provider::{validate_cursor_pagination, PgManager},
+    error::Error,
+    types::{
+        base64::Base64, move_function::MoveFunction, move_type::MoveType, object::Object,
+        sui_address::SuiAddress,
+    },
+};
+
+#[derive(Clone, Eq, PartialEq)]
+pub(crate) struct ProgrammableTransactionBlock(pub NativeProgrammableTransactionBlock);
+
+#[derive(Union, Clone, Eq, PartialEq)]
+enum TransactionInput {
+    OwnedOrImmutable(OwnedOrImmutable),
+    SharedInput(SharedInput),
+    Receiving(Receiving),
+    Pure(Pure),
+}
+
+/// A Move object, either immutable, or owned mutable.
+#[derive(SimpleObject, Clone, Eq, PartialEq)]
+#[graphql(complex)]
+struct OwnedOrImmutable {
+    address: SuiAddress,
+    version: u64,
+    digest: String,
+}
+
+/// A Move object that's shared.
+#[derive(SimpleObject, Clone, Eq, PartialEq)]
+struct SharedInput {
+    address: SuiAddress,
+    /// The version that this this object was shared at.
+    initial_shared_version: u64,
+    /// Controls whether the transaction block can reference the shared object as a mutable
+    /// reference or by value.
+    mutable: bool,
+}
+
+/// A Move object that can be received in this transaction.
+#[derive(SimpleObject, Clone, Eq, PartialEq)]
+#[graphql(complex)]
+struct Receiving {
+    address: SuiAddress,
+    version: u64,
+    digest: String,
+}
+
+/// BCS encoded primitive value (not an object or Move struct).
+#[derive(SimpleObject, Clone, Eq, PartialEq)]
+struct Pure {
+    /// BCS serialized and Base64 encoded primitive value.
+    bytes: Base64,
+}
+
+/// A single transaction, or command, in the programmable transaction block.
+#[derive(Union, Clone, Eq, PartialEq)]
+enum ProgrammableTransaction {
+    MoveCall(MoveCallTransaction),
+    TransferObjects(TransferObjectsTransaction),
+    SplitCoins(SplitCoinsTransaction),
+    MergeCoins(MergeCoinsTransaction),
+    Publish(PublishTransaction),
+    Upgrade(UpgradeTransaction),
+    MakeMoveVec(MakeMoveVecTransaction),
+}
+
+#[derive(Clone, Eq, PartialEq)]
+struct MoveCallTransaction(NativeMoveCallTransaction);
+
+/// Transfers `inputs` to `address`. All inputs must have the `store` ability (allows public
+/// transfer) and must not be previously immutable or shared.
+#[derive(SimpleObject, Clone, Eq, PartialEq)]
+struct TransferObjectsTransaction {
+    /// The objects to transfer.
+    inputs: Vec<TransactionArgument>,
+
```

Comment:
nit: extra space

Reply:
I see all of the transactions are like this, not sure if there's a specific reason for this formatting.

Reply:
The extra space is intentional, because each field has a doc comment, so having an extra space helps the eye associate the comment with its field.